// expo-cli/build/commands/init.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _keys;

function _load_keys() {
  return _keys = _interopRequireDefault(require('babel-runtime/core-js/object/keys'));
}

var _getIterator2;

function _load_getIterator() {
  return _getIterator2 = _interopRequireDefault(require('babel-runtime/core-js/get-iterator'));
}

var _stringify;

function _load_stringify() {
  return _stringify = _interopRequireDefault(require('babel-runtime/core-js/json/stringify'));
}

var _regenerator;

function _load_regenerator() {
  return _regenerator = _interopRequireDefault(require('babel-runtime/regenerator'));
}

var _asyncToGenerator2;

function _load_asyncToGenerator() {
  return _asyncToGenerator2 = _interopRequireDefault(require('babel-runtime/helpers/asyncToGenerator'));
}

var action = function () {
  var _ref = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee(projectDir, options) {
    var parentDir, dirName, root, validationResult, templateSpec, _ref2, workflow, initialConfig, packageManager, projectPath, cdPath;

    return (_regenerator || _load_regenerator()).default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            parentDir = void 0;
            dirName = void 0;

            if (!projectDir) {
              _context.next = 11;
              break;
            }

            root = _path.default.resolve(projectDir);

            parentDir = _path.default.dirname(root);
            dirName = _path.default.basename(root);
            validationResult = validateName(parentDir, dirName);

            if (!(validationResult !== true)) {
              _context.next = 9;
              break;
            }

            throw new (_CommandError || _load_CommandError()).default('INVALID_PROJECT_DIR', validationResult);

          case 9:
            _context.next = 16;
            break;

          case 11:
            if (!(options.parent && options.parent.nonInteractive)) {
              _context.next = 15;
              break;
            }

            throw new (_CommandError || _load_CommandError()).default('NON_INTERACTIVE', 'The project dir argument is required in non-interactive mode.');

          case 15:
            parentDir = process.cwd();

          case 16:
            templateSpec = void 0;

            if (!options.template) {
              _context.next = 22;
              break;
            }

            templateSpec = (0, (_npmPackageArg || _load_npmPackageArg()).default)(options.template);

            // For backwards compatibility, 'blank' and 'tabs' are aliases for
            // 'expo-template-blank' and 'expo-template-tabs', respectively.
            if ((templateSpec.name === 'blank' || templateSpec.name === 'tabs') && templateSpec.registry) {
              templateSpec.name = templateSpec.escapedName = 'expo-template-' + templateSpec.name;
            }
            _context.next = 26;
            break;

          case 22:
            _context.next = 24;
            return (0, (_prompt || _load_prompt()).default)({
              type: 'list',
              name: 'templateSpec',
              message: 'Choose a template:',
              choices: FEATURED_TEMPLATES.map(function (template) {
                return {
                  value: template.name,
                  name: (_chalk || _load_chalk()).default.bold(template.shortName) + '\n' + (0, (_wordwrap || _load_wordwrap()).default)(2, process.stdout.columns || 80)(template.description),
                  short: template.name
                };
              })
            }, {
              nonInteractiveHelp: '--template: argument is required in non-interactive mode. Valid choices are: ' + FEATURED_TEMPLATES.map(function (template) {
                return '\'' + template.shortName + '\'';
              }).join(', ') + ' or any custom template (name of npm package).'
            });

          case 24:
            _ref2 = _context.sent;
            templateSpec = _ref2.templateSpec;

          case 26:
            workflow = void 0;

            if (!options.workflow) {
              _context.next = 35;
              break;
            }

            if (!(options.workflow === 'managed' || options.workflow === 'advanced')) {
              _context.next = 32;
              break;
            }

            workflow = options.workflow;
            _context.next = 33;
            break;

          case 32:
            throw new (_CommandError || _load_CommandError()).default('BAD_ARGS', 'Invalid --workflow: \'' + options.workflow + '\'. Valid choices are: managed, advanced');

          case 33:
            _context.next = 36;
            break;

          case 35:
            workflow = 'managed';

          case 36:
            _context.next = 38;
            return promptForInitialConfig(parentDir, dirName, workflow, options);

          case 38:
            initialConfig = _context.sent;
            packageManager = void 0;

            if (!options.yarn) {
              _context.next = 44;
              break;
            }

            packageManager = 'yarn';
            _context.next = 56;
            break;

          case 44:
            if (!options.npm) {
              _context.next = 48;
              break;
            }

            packageManager = 'npm';
            _context.next = 56;
            break;

          case 48:
            _context.next = 50;
            return shouldUseYarnAsync();

          case 50:
            if (!_context.sent) {
              _context.next = 54;
              break;
            }

            _context.t0 = 'yarn';
            _context.next = 55;
            break;

          case 54:
            _context.t0 = 'npm';

          case 55:
            packageManager = _context.t0;

          case 56:
            _context.next = 58;
            return downloadAndExtractTemplate(templateSpec, parentDir, dirName, packageManager, workflow, initialConfig);

          case 58:
            projectPath = _context.sent;
            cdPath = _path.default.relative(process.cwd(), projectPath);

            if (cdPath.length > projectPath.length) {
              cdPath = projectPath;
            }
            (_log || _load_log()).default.nested('\nYour project is ready at ' + projectPath);
            (_log || _load_log()).default.nested('To get started, you can type:\n');
            if (cdPath) {
              // empty string if project was created in current directory
              (_log || _load_log()).default.nested('  cd ' + cdPath);
            }
            (_log || _load_log()).default.nested('  ' + packageManager + ' start\n');

          case 65:
          case 'end':
            return _context.stop();
        }
      }
    }, _callee, this);
  }));

  return function action(_x, _x2) {
    return _ref.apply(this, arguments);
  };
}();

var downloadAndExtractTemplate = function () {
  var _ref3 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee2(templateSpec, parentDir, dirName, packageManager, workflow, initialConfig) {
    return (_regenerator || _load_regenerator()).default.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            return _context2.abrupt('return', (_xdl || _load_xdl()).Exp.extractTemplateApp(templateSpec, _path.default.join(parentDir, dirName || initialConfig.slug), packageManager, workflow, initialConfig));

          case 1:
          case 'end':
            return _context2.stop();
        }
      }
    }, _callee2, this);
  }));

  return function downloadAndExtractTemplate(_x3, _x4, _x5, _x6, _x7, _x8) {
    return _ref3.apply(this, arguments);
  };
}();

var shouldUseYarnAsync = function () {
  var _ref4 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee3() {
    var version, answer;
    return (_regenerator || _load_regenerator()).default.wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            _context3.prev = 0;
            _context3.next = 3;
            return (0, (_spawnAsync || _load_spawnAsync()).default)('yarnpkg', ['--version']);

          case 3:
            version = _context3.sent.stdout.trim();

            if ((_semver || _load_semver()).default.valid(version)) {
              _context3.next = 6;
              break;
            }

            return _context3.abrupt('return', false);

          case 6:
            _context3.next = 8;
            return (0, (_prompt || _load_prompt()).default)({
              type: 'confirm',
              name: 'useYarn',
              message: 'Yarn v' + version + ' found. Use Yarn to install dependencies?'
            }, {
              nonInteractiveHelp: 'Please specify either --npm or --yarn to choose the installation method.'
            });

          case 8:
            answer = _context3.sent;
            return _context3.abrupt('return', answer.useYarn);

          case 12:
            _context3.prev = 12;
            _context3.t0 = _context3['catch'](0);
            return _context3.abrupt('return', false);

          case 15:
          case 'end':
            return _context3.stop();
        }
      }
    }, _callee3, this, [[0, 12]]);
  }));

  return function shouldUseYarnAsync() {
    return _ref4.apply(this, arguments);
  };
}();

var promptForInitialConfig = function () {
  var _ref5 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee4(parentDir, dirName, workflow, options) {
    var _config, template, _ref6, values, config, _iteratorNormalCompletion, _didIteratorError, _iteratorError, _iterator, _step, key;

    return (_regenerator || _load_regenerator()).default.wrap(function _callee4$(_context4) {
      while (1) {
        switch (_context4.prev = _context4.next) {
          case 0:
            if (!(options.parent && options.parent.nonInteractive)) {
              _context4.next = 27;
              break;
            }

            _config = {
              slug: dirName
            };

            if (!(!(0, (_isString || _load_isString()).default)(options.name) || options.name === '')) {
              _context4.next = 6;
              break;
            }

            throw new (_CommandError || _load_CommandError()).default('NON_INTERACTIVE', '--name: argument is required in non-interactive mode.');

          case 6:
            _config.name = options.name;

          case 7:
            if (!(workflow === 'advanced')) {
              _context4.next = 26;
              break;
            }

            if (!(!(0, (_isString || _load_isString()).default)(options.androidPackage) || options.androidPackage === '')) {
              _context4.next = 12;
              break;
            }

            throw new (_CommandError || _load_CommandError()).default('NON_INTERACTIVE', '--android-package: argument is required in non-interactive mode.');

          case 12:
            if (!(validateAndroidPackage(options.androidPackage) !== true)) {
              _context4.next = 16;
              break;
            }

            throw new (_CommandError || _load_CommandError()).default('INVALID_ARGUMENT', '--android-package: ' + validateAndroidPackage(options.androidPackage));

          case 16:
            _config.android = { package: options.androidPackage };

          case 17:
            if (!(!(0, (_isString || _load_isString()).default)(options.iosBundleIdentifier) || options.iosBundleIdentifier === '')) {
              _context4.next = 21;
              break;
            }

            throw new (_CommandError || _load_CommandError()).default('NON_INTERACTIVE', '--ios-bundle-identifier: argument is required in non-interactive mode.');

          case 21:
            if (!(validateIosBundleIdentifier(options.iosBundleIdentifier) !== true)) {
              _context4.next = 25;
              break;
            }

            throw new (_CommandError || _load_CommandError()).default('INVALID_ARGUMENT', '--ios-bundle-identifier: ' + validateIosBundleIdentifier(options.iosBundleIdentifier));

          case 25:
            _config.ios = { bundleIdentifier: options.iosBundleIdentifier };

          case 26:
            return _context4.abrupt('return', _config);

          case 27:
            template = {
              expo: {
                name: '{{name}}',
                slug: '{{slug}}'
              }
            };


            if (workflow === 'advanced') {
              template.android = {
                package: '{{android.package}}'
              };
              template.ios = {
                bundleIdentifier: '{{ios.bundleIdentifier}}'
              };
            }

            _context4.next = 31;
            return new (_enquirer || _load_enquirer()).Snippet({
              name: 'expo',
              message: 'Please enter a few initial configuration values.\n  Read more: https://docs.expo.io/versions/latest/workflow/configuration/',
              required: true,
              fields: [{
                name: 'name',
                message: 'The name of your app visible on the home screen',
                initial: (0, (_isString || _load_isString()).default)(options.name) ? options.name : undefined,
                filter: function filter(name) {
                  return name.trim();
                },
                required: true
              }, {
                name: 'slug',
                message: 'A URL friendly name for your app',
                initial: dirName,
                filter: function filter(name) {
                  return name.trim();
                },
                validate: function validate(name) {
                  return validateName(parentDir, name);
                },
                required: true
              }, {
                name: 'android.package',
                message: 'The package name for your Android app',
                initial: options.androidPackage,
                validate: validateAndroidPackage,
                required: true
              }, {
                name: 'ios.bundleIdentifier',
                message: 'The bundle identifier for your iOS app',
                initial: options.iosBundleIdentifier,
                validate: validateIosBundleIdentifier,
                required: true
              }],
              initial: 'slug',
              template: (0, (_stringify || _load_stringify()).default)(template, null, 2)
            }).run();

          case 31:
            _ref6 = _context4.sent;
            values = _ref6.values;
            config = {};
            _iteratorNormalCompletion = true;
            _didIteratorError = false;
            _iteratorError = undefined;
            _context4.prev = 37;

            for (_iterator = (0, (_getIterator2 || _load_getIterator()).default)((0, (_keys || _load_keys()).default)(values)); !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
              key = _step.value;

              (0, (_set || _load_set()).default)(config, key, values[key]);
            }
            _context4.next = 45;
            break;

          case 41:
            _context4.prev = 41;
            _context4.t0 = _context4['catch'](37);
            _didIteratorError = true;
            _iteratorError = _context4.t0;

          case 45:
            _context4.prev = 45;
            _context4.prev = 46;

            if (!_iteratorNormalCompletion && _iterator.return) {
              _iterator.return();
            }

          case 48:
            _context4.prev = 48;

            if (!_didIteratorError) {
              _context4.next = 51;
              break;
            }

            throw _iteratorError;

          case 51:
            return _context4.finish(48);

          case 52:
            return _context4.finish(45);

          case 53:
            return _context4.abrupt('return', config);

          case 54:
          case 'end':
            return _context4.stop();
        }
      }
    }, _callee4, this, [[37, 41, 45, 53], [46,, 48, 52]]);
  }));

  return function promptForInitialConfig(_x9, _x10, _x11, _x12) {
    return _ref5.apply(this, arguments);
  };
}();

var _chalk;

function _load_chalk() {
  return _chalk = _interopRequireDefault(require('chalk'));
}

var _fs = _interopRequireDefault(require('fs'));

var _xdl;

function _load_xdl() {
  return _xdl = require('xdl');
}

var _isString;

function _load_isString() {
  return _isString = _interopRequireDefault(require('lodash/isString'));
}

var _enquirer;

function _load_enquirer() {
  return _enquirer = require('enquirer');
}

var _semver;

function _load_semver() {
  return _semver = _interopRequireDefault(require('semver'));
}

var _set;

function _load_set() {
  return _set = _interopRequireDefault(require('lodash/set'));
}

var _spawnAsync;

function _load_spawnAsync() {
  return _spawnAsync = _interopRequireDefault(require('@expo/spawn-async'));
}

var _npmPackageArg;

function _load_npmPackageArg() {
  return _npmPackageArg = _interopRequireDefault(require('npm-package-arg'));
}

var _wordwrap;

function _load_wordwrap() {
  return _wordwrap = _interopRequireDefault(require('wordwrap'));
}

var _prompt;

function _load_prompt() {
  return _prompt = _interopRequireDefault(require('../prompt'));
}

var _log;

function _load_log() {
  return _log = _interopRequireDefault(require('../log'));
}

var _CommandError;

function _load_CommandError() {
  return _CommandError = _interopRequireDefault(require('../CommandError'));
}

var _path = _interopRequireDefault(require('path'));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var FEATURED_TEMPLATES = [{
  shortName: 'blank',
  name: 'expo-template-blank',
  description: 'minimum dependencies to run and an empty root component'
}, {
  shortName: 'tabs',
  name: 'expo-template-tabs',
  description: 'several example screens and tabs using react-navigation'
}];

var _downloadIsSlowPrompt = false;

function validateName(parentDir, name) {
  if (!/^[a-z0-9@.\-_]+$/i.test(name)) {
    return 'The project name can only contain URL-friendly characters.';
  }
  if (typeof name !== 'string' || name === '') {
    return 'The project name can not be empty.';
  }
  var dir = _path.default.join(parentDir, name);
  if (!isNonExistentOrEmptyDir(dir)) {
    return 'The path "' + dir + '" already exists. Please choose a different parent directory or project name.';
  }
  return true;
}

function isNonExistentOrEmptyDir(dir) {
  try {
    return _fs.default.statSync(dir).isDirectory() && _fs.default.readdirSync(dir).length === 0;
  } catch (error) {
    if (error.code === 'ENOENT') {
      return true;
    }
    throw error;
  }
}

function validateAndroidPackage(value) {
  if (!(0, (_isString || _load_isString()).default)(value) || value === '') {
    return 'Android package identifier must not be empty.';
  }
  return (/^[a-zA-Z][a-zA-Z0-9\_]*(\.[a-zA-Z][a-zA-Z0-9\_]*)+$/.test(value) || "Only alphanumeric characters, '.' and '_' are allowed, and each '.' must be followed by a letter."
  );
}

function validateIosBundleIdentifier(value) {
  if (!(0, (_isString || _load_isString()).default)(value) || value === '') {
    return 'iOS bundle identifier must not be empty.';
  }
  return (/^[a-zA-Z][a-zA-Z0-9\-\.]+$/.test(value) || "Must start with a letter and only alphanumeric characters, '.' and '-' are allowed."
  );
}

exports.default = function (program) {
  program.command('init [project-dir]').alias('i').description('Initializes a directory with an example project. Run it without any options and you will be prompted for the name and type.').option('-t, --template [name]', 'Specify which template to use. Valid options are "blank", "tabs" or any npm package that includes an Expo project template.').option('--npm', 'Use npm to install dependencies. (default when Yarn is not installed)').option('--yarn', 'Use Yarn to install dependencies. (default when Yarn is installed)').option('--workflow [name]', 'The workflow to use. managed (default) or advanced').option('--name [name]', 'The name of your app visible on the home screen.').option('--android-package [name]', 'The package name for your Android app.').option('--ios-bundle-identifier [name]', 'The bundle identifier for your iOS app.').asyncAction(action);
};

module.exports = exports['default'];
